lib/Onyx.js

```diff
@@ -768,11 +768,12 @@ function mergeCollection(collectionKey, collection) {
                 promises.push(Storage.multiSet(keyValuePairsForNewCollection));
             }
 
-            // Merge original data to cache
-            cache.merge(collection);
-
-            // Optimistically inform collection subscribers
-            keysChanged(collectionKey, collection);
+            // Prefill cache if necessary by calling get() on any existing keys and then merge original data to cache
+            // and update all subscribers
+            Promise.all(_.map(existingKeys, get)).then(() => {
+                cache.merge(collection);
+                keysChanged(collectionKey, collection);
+            });
```

Comment:
>if get has to read from disk, data might not yet be written to disk (when write is still pending), and still result in partial data being added to cache

If you can, please explain how exactly it can result in partial data being added to the cache? I would love to write a test for it and we can come up with a solution if it's possible, but I want to make sure there we understand how it can happen.

>delaying cache update opens a door for problems - other get/connect calls might use stale cache information

One alternative I can think of is to prefill the cache.

>it might delay the UI update a bit, since now keysChanged would be delayed if get actually has to read something

Yes, but better for something to be delayed than send partial data without previous data initially in storage.

>The goal with cache is to update it sync as fast as possible to avoid concurrency issues

I think I understand this argument, but I'm not sure that's how `merge()` works today. Here's the implementation for `merge()` which also potentially blocks with a `get()` call

https://github.com/Expensify/react-native-onyx/blob/032ff54b4374c93ee12676792a6e029c61ea1e06/lib/Onyx.js#L672-L681

`set()` is the only thing which modifies the data in the cache first before reading existing values 

https://github.com/Expensify/react-native-onyx/blob/032ff54b4374c93ee12676792a6e029c61ea1e06/lib/Onyx.js#L548-L551

>Perhaps you'd like this alternative:
>Keep code here unchanged
>Update code in cache.merge to exclude merging any keys that are not currently present in cache

I do like this idea and it solves this problem in a different way, but we would defer the filling of the cache. I'm not sure if that is bad or good or doesn't make a difference and would be willing to try it if it sounds simpler. Thanks!

Reply:
Ok so pulling on that last thread and I can't get it working. It leads to a stale cache that is missing the data. Here are the changes I made:

- Reverted the change here.
- Added this so we would not set data if no existing key was in the cache

```diff
@@ -107,11 +110,19 @@ class OnyxCache {
     }

     /**
-     * Deep merge data to cache, any non existing keys will be created
+     * Deep merge data to cache, any non existing keys will not be merged
      * @param {Record<string, *>} data - a map of (cache) key - values
      */
     merge(data) {
-        this.storageMap = lodashMerge({}, this.storageMap, data);
+        // Filter any keys we have not yet cached
+        const newData = _.reduce(data, (prev, curr, key) => {
+            if (this.hasCacheForKey(key)) {
+                return {...prev, [key]: curr};
+            }
+            return prev;
+        }, {});
+
+        this.storageMap = lodashMerge({}, this.storageMap, newData);

         const storageKeys = this.getAllKeys();
```
- Re-ran tests and added a subscriber (which should trigger the cache fill)
- Storage ends up correct, but cache is wrong 

I think it's because the subscriber can be added at anytime while we are adding new data. We already skipped the cache for those changes and yes they are making it to storage correctly. But if you add a subscriber at the same time it will only get the result of what has been written and save that to the cache.

Reply:
Hey @kidroca, I couldn't figure this one out, but got your other idea working. Let me know if you have any other thoughts here.

Reply:
I think prefilling cache might be our safest and optimal option here

It seems that for Native:
1. Calling AsyncStorage.get
2. Then calling AsyncStorage.set, before the get is completed
3. `get` would resolve with the value from step 2)

I'm not certain this would work out the same way on Web/Desktop
It used to work that way when we used `window.localStorage` because local storage calls are sync and happen sequentially

So we're back to: "Oh, Cache save us from concurrency issues" 😄 
And to do that it might be best to leverage `multiGet` and get everything but safe eviction keys in cache during init
Then we don't have to delay or change logic that updates cache and cache should never become stale

---

Another strategy could be to recreated the above 3 steps ourselves 
- capture pending writes (we already have a queue)
- when a read happens check whether we're not writing or are about to write the requested key
- if we are - resolve the get with the value from the write queue, otherwise read from disk